src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -604,6 +604,38 @@ def _visit_max_over(self, node: foast.Call, **kwargs) -> foast.Call:
     def _visit_min_over(self, node: foast.Call, **kwargs) -> foast.Call:
         return self._visit_reduction(node, **kwargs)
 
+    def _visit_astype(self, node: foast.Call, **kwargs) -> foast.Call:
+        casted_obj_type = node.args[0].type
+        dtype_obj_type = node.args[1].type
+        if not isinstance(dtype_obj_type, ct.FunctionType) or not (
+            type_info.is_arithmetic(dtype_obj_type.returns)
+            or type_info.is_logical(dtype_obj_type.returns)
+        ):
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Incompatible argument in call to `{node.func.id}`'s new dtype. "
+                f"Expected an arithmetic or boolean dtype, but got {dtype_obj_type}.",
+            )
```

Comment:
Please document why it makes sense that `dtype_obj_type` is a FunctionType. I guess it is because `float32` etc are all functions and we abuse it's return type.

Where do we actually deduce the type of these? @tehrengruber